### PR TITLE
Global parameters for the different number of hydro/mhd variables

### PR DESCRIPTION
**Pull request by tinecolman**
Introduce `nvar_all`, `neul` and `nhydro` variables to simplify differences between MHD versus HYDRO. This allows for more general code and less preprocessor statements.

### CODE REVIEW

File: mhd/hydro_parameters.f90 [hunk 1]
```diff
@@ -5,16 +5,30 @@ module hydro_parameters
   use amr_parameters
 
   ! Number of independant variables
+
+  ! Euler variables: density, velocity, pressure
+  ! always 3 velocities when MHD
+  integer,parameter::neul=5
```
Comment (nbrucy) [inline mhd/hydro_parameters.f90:12]: Could neul be named neuler instead?

File: mhd/hydro_parameters.f90 [hunk 2]
```diff
@@ -5,16 +5,30 @@ module hydro_parameters
...
+#else
+  integer,parameter::nhydro=NHYDRO
+#endif
+  ! non-thermal energies
 #ifndef NENER
   integer,parameter::nener=0
 #else
   integer,parameter::nener=NENER
 #endif
+  ! total amount of variables
```
Comment (nbrucy) [inline mhd/hydro_parameters.f90:25]: total amount of independent variables? (to distinguish from nvar_all)

File: mhd/hydro_parameters.f90 [hunk 3]
```diff
@@ -32,59 +46,47 @@ module hydro_parameters
...
-  real(dp)::floor_d=1d-10   ! Density floor
-  real(dp)::floor_u=1d-10   ! Velocity floor
-  real(dp)::floor_p=1d-10   ! Pressure floor
-  real(dp)::floor_A=1d-10   ! Bx floor
-  real(dp)::floor_B=1d-10   ! By floor
-  real(dp)::floor_C=1d-10   ! Bz floor
-  real(dp)::floor_b2=1d-10  ! B L2 norm floor
-  real(dp)::mass_sph=0.0D0   ! mass_sph
-#if NENER>0
-  real(dp),dimension(1:NENER)::err_grad_prad=-1.0
```
Comment (nbrucy) [inline mhd/hydro_parameters.f90:67]: Where are these gone?

File: mhd/output_hydro.f90 [hunk 4]
```diff
@@ -110,41 +111,44 @@ subroutine backup_hydro(filename, filename_desc)
...
                  call generic_dump(field_name, info_var_count, xdp, unit_out, dump_info_flag, unit_info)
               end do
 #endif
-              do i = 1, ncache ! Write thermal pressure
+              ! Write thermal pressure
+              do i = 1, ncache
                  d = max(uold(ind_grid(i)+iskip, 1), smallr)
-                 u = uold(ind_grid(i)+iskip, 2)/d
-                 v = uold(ind_grid(i)+iskip, 3)/d
-                 w = uold(ind_grid(i)+iskip, 4)/d
```
Comment (nbrucy) [inline mhd/output_hydro.f90:125]: Why are these changes relevant for the PR?

**Review by nbrucy — CHANGES_REQUESTED** [review]
This makes the counting of variable easier to understand, with no consequences for end user (maybe a very small performance loss due to moving some computation at runtime?)

Maybe this can be further improved by adding some developer documentation in `doc/dev_docs` explaining what are the different number of variables `nvar`, `nhydro`, `neul` (why not `neuler` btw?), `nvar_all`, `nener`, etc ... and how they can be used.

File: mhd/hydro_parameters.f90 (hunk 1 above)
Comment (tinecolman) [reply mhd/hydro_parameters.f90:12]: I used neul because it was already used in some places in the code. I simple extended its use.

Comment (nbrucy) [reply mhd/hydro_parameters.f90:12]: Should we keep it like this?

File: mhd/output_hydro.f90 (hunk 4 above)
Comment (tinecolman) [reply mhd/output_hydro.f90:125]: Here I've synchronized the hydro and mhd version of the file. Reducing the amount of duplicate code between hydro and mhd was one of the motivations for introducing these new variables.

File: hydro/hydro_parameters.f90 [hunk 5]
```diff
@@ -6,16 +6,28 @@ module hydro_parameters
...
   integer,parameter::nener=NENER
 #endif
+  ! total amount of variables
 #ifndef NVAR
-  integer,parameter::nvar=ndim+2+nener
+  integer,parameter::nvar=nhydro+nener
 #else
   integer,parameter::nvar=NVAR
 #endif
+  integer,parameter::nvar_all=nvar
```
Comment (cphyc) [inline hydro/hydro_parameters.f90:29]: Regarding @nbrucy's comment about performance, all the variables are defined as compile-time, so the compiler should optimise them as such. If truly this is a worry, one could also replace these definitions with compiler directives, i.e. in the Makefile, create `DEFINES = -DNVAR_all=[...] -Dnhydro=[...]` etc. and used capitalized form to distinguish from runtime params.

File: mhd/hydro_parameters.f90 (hunk 3 above)
Comment (tinecolman) [reply mhd/hydro_parameters.f90:67]: You mean err_grad_prad and err_grad_var? They are unused variables, so I removed them.

File: mhd/hydro_parameters.f90 (hunk 1 above)
Comment (tinecolman) [reply mhd/hydro_parameters.f90:12]: In my opinion, neuler is not particulary more clear than neul. In any case, it is explained where the variable is defined. So I would opt for keeping neul.

File: mhd/hydro_parameters.f90 (hunk 3 above)
Comment (rteyssier) [reply mhd/hydro_parameters.f90:67]: They are unused but they could be used in the future or in a patch.
I would rather keep them.

**Review by rteyssier — CHANGES_REQUESTED** [review]
I would rather keep the radiation and passive scalar initial conditions and boundary conditions variables.

File: hydro/hydro_parameters.f90 (hunk 5 above)
Comment (tinecolman) [reply hydro/hydro_parameters.f90:29]: For nhydro it is there -DNHYDRO=$(NHYDRO). Should I add the others? I would think it is not an issue, but we don't really have performance tests in place that would cover the different configurations of these parameters so I can't know for sure.

File: mhd/hydro_parameters.f90 (hunk 3 above)
Comment (tinecolman) [reply mhd/hydro_parameters.f90:67]: I put them back.

File: mhd/hydro_parameters.f90 (hunk 2 above)
Comment (tinecolman) [reply mhd/hydro_parameters.f90:25]: Both nvar and nvar_all are total number of variables in a way...

**Comment by tinecolman** [comment]
I've added some dev doc. Can be the start of a more detailed documentation on some implementation aspects.

**Review by rteyssier — APPROVED** [review]

**Review by nbrucy — APPROVED** [review]
I added the new doc page in the index and corrected some typos (see https://ramses-organisation--36.org.readthedocs.build/en/36/). Good for me!

**Pull request merged**